tools/xmlschema.py

```diff
@@ -0,0 +1,312 @@
+    lines = []
+
+    elem_name = get_attribute(element, "name")
+    elem_type = get_attribute(element, "type")
+
+    elements = element.findall("element")
+    attributes = element.findall("attribute")
+    includes = element.findall("include")
+
+    lines.extend(print_documentation(element))
+    lines.append(
+        "<xsd:include schemaLocation='http://sdformat.org/schemas/types.xsd'/>"
+    )
+
+    # Reference any includes in the SDF file
+    for include in includes:
+        lines.extend(print_include(include))
+
+    if len(elements) or len(attributes) or len(includes):
+        lines.append(f"<xsd:element name='{elem_name}'>")
+        lines.append("  <xsd:complexType>")
+
+        if elem_name != "plugin" and (len(elements) or len(includes)):
+            lines.append("    <xsd:choice maxOccurs='unbounded'>")
+
+        for child_element in elements:
+            if "copy_data" in child_element.attrib:
+                element_lines = print_plugin_element(child_element)
+                lines.extend(indent_lines(element_lines, 4))
+            else:
+                element_lines = print_element(child_element)
+                lines.extend(indent_lines(element_lines, 6))
+
+        for include_element in includes:
+            element_lines = print_include_ref(include_element, sdf_root_dir)
+            lines.extend(indent_lines(element_lines, 6))
+
+        if elem_name != "plugin" and (len(elements) or len(includes)):
+            lines.append("    </xsd:choice>")
+
+        for attribute_element in attributes:
+            lines.extend(indent_lines(print_attribute(attribute_element), 4))
+
+        lines.append("  </xsd:complexType>")
+        lines.append("</xsd:element>")
+    else:
+        if elem_type and is_std_type(elem_type):
+            elem_type = xsd_type_string(elem_type)
+        else:
+            elem_type = ""
+
+        lines.append(f"<xsd:element name='{elem_name}'{elem_type} />")
+    return lines
+
+
+def process(input_file_sdf: str, sdf_dir: str) -> List[str]:
+    """
+    Produce an XSD file from an input SDF file
+    """
+    lines = []
+    tree = ElementTree.parse(input_file_sdf)
+    root = tree.getroot()
+    lines.append("<?xml version='1.0' encoding='UTF-8'?>")
+    lines.append("<xsd:schema xmlns:xsd='http://www.w3.org/2001/XMLSchema'>")
+    lines.extend(indent_lines(print_xsd(root, sdf_dir), 2))
+    lines.append("</xsd:schema>")
+    return lines
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser("xmlschema.py")
+    parser.add_argument("--input-file")
+    parser.add_argument("--sdf-dir")
+    parser.add_argument("--output-dir")
+    args = parser.parse_args()
+
+    input_file = os.path.abspath(args.input_file)
+    output_lines = process(input_file, args.sdf_dir)
+    fname = os.path.splitext(os.path.basename(args.input_file))[0]
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    output_file = os.path.join(args.output_dir, f"{fname}.xsd")
```

Comment:
Just an observation: the ruby script handles multiple top-level `<element>` tags 
while this doesn't, but I don't see any `.sdf` description files with multiple 
`<element>` tags.

```
xmlstarlet sel -t -v "count(/element)" sdf/**/*.sdf

```
Also, I believe xml allows only one top-level element.

Reply:
Agreed